packages/doenetml-worker-rust/lib-doenetml-core/src/core/dast/flat_dast/untagged_flat_dast_merge.rs

```diff
@@ -0,0 +1,229 @@
+        parent: Option<Index>,
+    ) -> UntaggedContent {
+        let ret = match node {
+            DastElementContent::Text(_) => {
+                // A text node doesn't get pushed to `nodes`, so the value of `idx` doesn't matter.
+                // We pass in `Index::MAX` so that an error will be thrown if this value is used.
+                self.update_content(node, Index::MAX, parent)
+            }
+            _ => {
+                // We push an error onto the nodes stack and then we do an in-place update of
+                // that error to the correct node. This avoids code duplication.
+                let idx = self.nodes.len();
+                self.nodes.push(FlatNode::Error(FlatError::with_message(
+                    "TEMPORARY NODE CREATED DURING `merge_content()`".to_string(),
+                    idx,
+                )));
+                self.update_content(node, idx, parent)
+            }
+        };
+
+        if parent.is_none() {
+            self.children.push(ret.clone());
+        }
+        ret
+    }
+
+    /// Set the children of an element node.
+    pub fn set_children(&mut self, idx: usize, children: Vec<UntaggedContent>) {
+        match &mut self.nodes[idx] {
+            FlatNode::Element(elm) => {
+                elm.children = children;
+            }
+            _ => panic!("set_children called on non-element node"),
+        }
+    }
+
+    /// Set the attributes of an element node.
+    fn set_attributes(&mut self, idx: usize, attributes: Vec<FlatAttribute>) {
+        match &mut self.nodes[idx] {
+            FlatNode::Element(elm) => {
+                elm.attributes = attributes;
+            }
+            _ => panic!("set_attributes called on non-element node"),
+        }
+    }
+
+    /// Set the input of a function macro node.
+    fn set_function_macro_input(&mut self, idx: usize, args: Vec<Vec<UntaggedContent>>) {
+        match &mut self.nodes[idx] {
+            FlatNode::FunctionMacro(func_macro) => {
+                func_macro.input = Some(args);
+            }
+            _ => panic!("set_function_macro_args called on non-function-macro node"),
+        }
+    }
+
+    /// Set the node at position `idx` to be the specified element.
+    /// The element will be initialized with empty children and attributes.
+    fn set_element(&mut self, node: &DastElement, idx: Index, parent: Option<Index>) -> usize {
+        self.nodes[idx] = FlatNode::Element(FlatElement {
+            name: node.name.clone(),
+            children: Vec::new(),
+            attributes: Vec::new(),
+            position: node.position.clone(),
+            parent,
+            idx,
+            // It is impossible to directly set `extending` in DAST; it is computed later.
+            extending: None,
+        });
+        idx
+    }
+
+    /// Set the node at position `idx` to be the specified error.
+    fn set_error(&mut self, node: &DastError, idx: Index, parent: Option<Index>) -> usize {
```

Comment:
It's a little confusing to use "position" to describe `idx`, when "position" means something else in the function.